File: packages/backend-shared/src/health/health-check.service.ts
import type { DependencyHealth, HealthResult, ReadinessResult } from '../types';
import type { HealthIndicator } from './types';

export interface HealthCheckServiceOptions {
  service: string;
  version: string;
  /**
   * Override start time (ms since epoch) for deterministic tests.
   */
  startTimeMs?: number;
}

/**
 * HealthCheckService
 *
 * - `/health` (liveness): basic process status
 * - `/ready` (readiness): checks registered dependency indicators
 */
export class HealthCheckService {
  private readonly indicators = new Map<string, HealthIndicator>();
  private readonly startTimeMs: number;

  constructor(private readonly options: HealthCheckServiceOptions) {
    this.startTimeMs = options.startTimeMs ?? Date.now();
  }

  registerIndicator(indicator: HealthIndicator): void {
    this.indicators.set(indicator.name, indicator);
  }

  listIndicators(): string[] {
    return Array.from(this.indicators.keys());
  }

  checkHealth(): HealthResult {
    return {
      status: 'ok',
      service: this.options.service,
      version: this.options.version,
      uptime: Math.floor((Date.now() - this.startTimeMs) / 1000),
      timestamp: new Date().toISOString(),
    };
  }

  async checkReadiness(): Promise<ReadinessResult> {
    const dependencyResults = await Promise.all(
      Array.from(this.indicators.values()).map(async (indicator) => {
        const start = Date.now();
        try {
          const status = await indicator.check();
          return {
            name: indicator.name,
            status: status.status,
            latencyMs: status.latencyMs ?? Date.now() - start,
            error: status.error,
          } satisfies DependencyHealth;
        } catch (error) {
          return {
            name: indicator.name,
            status: 'unhealthy',
            latencyMs: Date.now() - start,
            error: error instanceof Error ? error.message : String(error),
          } satisfies DependencyHealth;
        }
      }),
    );

    const allHealthy = dependencyResults.every((d) => d.status === 'healthy');
    return {
      ...this.checkHealth(),
      status: allHealthy ? 'ok' : 'error',
      dependencies: dependencyResults,
    };
  }
}

File: packages/backend-shared/python/superboard_shared/health/health_check_service.py
"""
Health Check Service (Python) - Placeholder for Task 1
Full implementation will be done in later tasks
"""

from typing import Dict, Any, List
from .types import HealthResult, ReadinessResult


class HealthCheckService:
    """
    Health Check Service class (placeholder)
    Full implementation will be done in Task 8
    """
    
    def __init__(self):
        self.indicators: Dict[str, Any] = {}
    
    def register_indicator(self, name: str, indicator: Any) -> None:
        """Register a health indicator (placeholder)"""
        self.indicators[name] = indicator
    
    async def check_health(self) -> HealthResult:
        """Check basic health (placeholder)"""
        return HealthResult(
            status='ok',
            service='placeholder',
            version='0.1.0',
            uptime=0,
            timestamp='2024-01-01T00:00:00Z'
        )
    
    async def check_readiness(self) -> ReadinessResult:
        """Check readiness with dependencies (placeholder)"""
        return ReadinessResult(
            status='ok',
            service='placeholder',
            version='0.1.0',
            uptime=0,
            timestamp='2024-01-01T00:00:00Z',
            dependencies=[]
        )